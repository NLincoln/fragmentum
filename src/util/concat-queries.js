export const concatQueries = (arr, joinStr = ", ") => {
  const reduced = arr.filter(f => f.query).reduce(
    (prev, curr) => ({
      query: prev.query.concat(curr.query),
      binds: prev.binds.concat(curr.binds)
    }),
    {
      query: [],
      binds: []
    }
  );
  return {
    query: reduced.query.join(joinStr),
    binds: reduced.binds
  };
};
