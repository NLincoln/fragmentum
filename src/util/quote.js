export default function quote(val, opts = {}) {
  if (val === "*") {
    return val;
  }
  if (opts.single) {
    return `'${val.replace(`'`, `\\'`)}'`;
  }
  if (opts.parens) {
    return `(${val})`;
  }
  return `"${val.replace(`"`, `\\"`)}"`;
}
