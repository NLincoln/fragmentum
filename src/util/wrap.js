export default Klass => (...args) => new Klass(...args);
