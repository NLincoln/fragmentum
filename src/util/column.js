import quote from "./quote";

export default raw =>
  raw
    .split(".")
    .map(quote)
    .join(".");
