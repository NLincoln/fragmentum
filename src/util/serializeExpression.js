import { Builder } from "../builder";
import Expression from "../expressions/expression";
import FromFragment from "../fragments/from";
import quote from "./quote";
import columnQuote from "./column-quote";

export const serializePair = ([key, alias]) => {
  if (key && alias) {
    return `${columnQuote(key)} AS ${quote(alias)}`;
  }
  return null;
};
const serializeObject = column =>
  Object.keys(column)
    .map(key => serializePair([key, column[key]]))
    .join(", ");

const serializeColumn = column => {
  if (Array.isArray(column)) {
    return serializePair(column);
  } else if (typeof column === "object") {
    return serializeObject(column);
  }
  return columnQuote(column);
};

export const serializeMaybeExpression = column => {
  // Expressions know how to serialize themselves, FromFragment indicates that
  // we have a wrapped subquery, so we just forward that too.
  if (column instanceof Expression || column instanceof FromFragment) {
    return column.serialize();
  } else if (column instanceof Builder) {
    let { query, binds } = column.serialize({ partial: true });
    return {
      query: `(${query}) AS ${quote(column.alias)}`,
      binds
    };
  }
  return {
    query: serializeColumn(column),
    binds: []
  };
};
