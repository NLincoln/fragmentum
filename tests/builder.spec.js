import { builder, select, from, where, ops, value } from "fragmentum";
import { testQuery } from "./util";

describe("builder", () => {
  testQuery(
    "concat-ing from another builder",
    builder(builder().select(), builder().from("users")),
    `SELECT * FROM "users";`
  );
  testQuery("list of fragments", builder(select("user"), from("users")));
  test("method chaining", () => {
    const sql = builder()
      .select("user")
      .from("users")
      .serialize();
    expect(sql).toMatchSnapshot();
  });
  test("a healthy mix", () => {
    expect(
      builder(select())
        .from("users")
        .serialize()
    ).toMatchSnapshot();
  });
});

describe("conditional fragments", () => {
  testQuery(
    "conditional fragments as part of a builder / concat",
    builder(false, select("users"))
      .from("users")
      .concat(false, select("id")),
    `SELECT "users", "id" FROM "users";`
  );
  testQuery(
    "conditional parts of an expression",
    builder(select(), from("users"), where(false)),
    `SELECT * FROM "users";`
  );
});
describe("Serializing just fragments", () => {
  testQuery(
    "omitting the select portion",
    builder(from("users")),
    `FROM "users"`
  );
  testQuery("omitting the from portion", builder(select()), "SELECT *");
  test("select", () => {
    expect(select().serialize()).toEqual("*");
  });
  testQuery("from", from("users"), `"users"`);
  test("where", () => {
    expect(where(ops.eq("username", value(2))).serialize().query).toEqual(
      `("username" = '2')`
    );
  });
});

testQuery(
  "table.column syntax",
  builder().select(
    "user.*",
    "user.username",
    "groups.groupname",
    "user.id",
    ["user.first_name", "first"],
    {
      "user.last_name": "last"
    }
  ),
  `SELECT "user".*, "user"."username", "groups"."groupname", "user"."id", "user"."first_name" AS "first", "user"."last_name" AS "last"`
);
describe.skip("where");
describe.skip("having");
describe.skip("group by");
describe.skip("limit/offset");
describe.skip("immutability");
describe.skip("custom SQL functions");
describe.skip("SQL functions");
