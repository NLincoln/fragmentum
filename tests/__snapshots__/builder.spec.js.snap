// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder a healthy mix 1`] = `
Object {
  "binds": Object {},
  "query": "SELECT * FROM \\"users\\";",
}
`;

exports[`builder list of fragments 1`] = `
Object {
  "binds": Object {},
  "query": "SELECT \\"user\\" FROM \\"users\\";",
}
`;

exports[`builder method chaining 1`] = `
Object {
  "binds": Object {},
  "query": "SELECT \\"user\\" FROM \\"users\\";",
}
`;
